sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+ feature-test macro always has this value.
+|===
+
+=== Control Flow
+
+The SYCL specification defines
+link:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#control-flow[control
+flow] as below:
+
+> When all work-items in a group are executing the same sequence of statements,
+> they are said to be executing under _converged_ control flow. Control flow
+> _diverges_ when different work-items in a group execute a different sequence of
+> statements, typically as a result of evaluating conditions differently (e.g.
+> in selection statements or loops).
+
+This extension introduces some new terminology to describe other kinds of
+control flow, to simplify the description of the behavior for new group types.
+
+A work-item is said to have _active_ control flow for a given point of
+execution if it follows the sequence of statements currently being executed.
+
+Two or more work-items are considered to have _simultaneous_ execution at a given
+point if they execute together. The amount of and granularity of simultaneous
+execution is implementation-defined.
+
+A _tangle_ is a collection of work-items from the same sub-group executing
+under converged control flow.
+
+A _simultaneous tangle_ is a subset of a tangle representing a set of
+work-items executing under converged control flow and with simultaneous
+execution.
+
+
+=== Group Taxonomy
+
+_Fixed topology groups_ are groups which represent the hierarchical execution
+model topology used by SYCL kernels. These groups are implicitly created by an
+implementation when a SYCL kernel function is enqueued. The following group
+types are fixed topology groups:
+
+- `root_group` (if sycl_ext_oneapi_root_group is supported)
+- `group`
+- `sub_group`
+
+_User-constructed groups_ are explicitly created by a developer (e.g. by
+partitioning one of the fixed topology groups). This extension introduces the
+following user-constructed groups:
+
+- `ballot_group`
+- `cluster_group`
+- `tangle_group`
+- `opportunistic_group`
+- `masked_sub_group`
+
+The `is_fixed_topology_group` and `is_user_constructed_group` traits can be
+used to detect whether a group type represents a fixed topology or
+user-constructed group, respectively.
+
+[source, c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <class T>
+  struct is_fixed_topology_group;
+
+  template <class T>
+  inline constexpr
+  bool is_fixed_topology_group_v = is_fixed_topology_group<T>::value;
+
+
+  template <class T>
+  struct is_user_constructed_group;
+
+  template <class T>
+  inline constexpr bool
+  is_user_constructed_group_v = is_user_constructed_group<T>::value;
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+`is_fixed_topology_group<T>::value` is `std::true_type` if `T` is one of:
+`root_group`, `group` or `sub_group`.
+
+`is_user_constructed_group<T>::value` is `std::true_type` if `T` is one of:
+`ballot_group`, `cluster_group`, `tangle_group`, `opportunisic_group` or
+`masked_sub_group`.
```

Comment:
This is a really good point. I'm now a little unsure about SYCL 2020's usage of `is_group<T>`.

My immediate thought was that these new groups shouldn't have `sycl::is_group<T>::value` set to `std::true_type`, because they might match against existing overloads of group functions and algorithms defined by SYCL 2020.  But SYCL 2020 just says that the functions will be "available only if" the type trait is true -- it _doesn't_ say that an implementation can't additionally specialize the functions (i.e. by providing overloads for specific groups).

I think that means that these groups _can_ return true here, because all that `is_group` currently controls is whether the group algorithms are implemented.  Do you agree?

Reply:
The core SYCL spec says that `is_group<T>` is a constraint for a number of the group functions and algorithms.  Are all of those functions and algorithms supported for the new user-defined groups in this extension?  If the answer is "yes", then I think the extension should state that `is_group<T>` is true for these user-defined groups.  This would be a formal way of stating that the core spec's group functions and algorithms are supported for these new user-defined groups.

Note also that the core spec says this:

> A SYCL implementation may introduce additional specializations of `is_­group<T>` for implementation-defined group types, if the interface of those types supports all member functions and static members common to the `group` and `sub_group` classes.

Is this true of the user-defined groups?  (I think it is, but just checking.)

Reply:
I think we support all the same functions and algorithms and have the same member functions and static members.  I've added some wording saying that `is_group<T>` is true for user-defined groups in 4207fd76a.